File: golang/cmd/image-scale/image-scale.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"image/jpeg"
	"io/fs"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"github.com/dasbd72/images-processing-benchmarks/golang/internal/minioimages"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
	"github.com/nfnt/resize"
)

const (
	localPath       = "storage/"                         // path of local storage
	endpoint        = "10.121.240.169:9000"              // without http://
	accessKeyID     = "LbtKL76UbWedONnd"                 //
	secretAccessKey = "Bt0Omfh0S3ud5VEQAVR85CwinSULl3Sj" // secret key from minio console
	useSSL          = false                              // no certificate
)

var (
	port int  // port of host
	dry  bool // dry run
)

func init() {
	flag.IntVar(&port, "port", 9090, "Port of server")
	flag.BoolVar(&dry, "dry", false, "Dryrun")
}

func main() {
	flag.Parse()

	log.Printf("Server port: %d\n", port)

	http.HandleFunc("/", handleRequest)
	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(port), nil))
}

func handleRequest(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case "GET":
		fmt.Fprintf(w, "Not For You\n")
	case "POST":
		err := r.ParseForm()
		if err != nil {
			fmt.Fprintf(w, "ParseForm() err: %v", err)
			return
		}

		var request struct {
			Bucket string // Bucket name
			Source string // Path of image directory
			Width  int
			Height int
		}
		err = json.NewDecoder(r.Body).Decode(&request)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		bucketName := filepath.Clean(request.Bucket)
		downloadPath := filepath.Clean(request.Source) + "/"
		uploadPath := filepath.Clean(request.Source) + "-" + time.Now().Format("20060102150405") + "/"

		connect_start := time.Now()
		// Initialize minio client object.
		minioClient, err := minio.New(endpoint, &minio.Options{
			Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
			Secure: useSSL,
		})
		if err != nil {
			log.Fatalln(err)
		} else {
			log.Printf("Connected to %s\n", endpoint)
		}
		minioimages.CreateBucket(minioClient, bucketName)
		connect_duration := time.Since(connect_start)

		download_start := time.Now()
		if !dry {
			minioimages.DownloadImages(minioClient, bucketName, downloadPath, localPath)
		}
		download_duration := time.Since(download_start)

		scale_start := time.Now()
		if !dry {
			scaleImages(localPath, request.Width, request.Height)
		}
		scale_duration := time.Since(scale_start)

		upload_start := time.Now()
		if !dry {
			minioimages.UploadImages(minioClient, bucketName, localPath, uploadPath)
		}
		upload_duration := time.Since(upload_start)

		log.Println("Connect:", connect_duration)
		log.Println("Download:", download_duration)
		log.Println("Scale:", scale_duration)
		log.Println("Upload:", upload_duration)

		fmt.Fprintf(w, "Connect: %s\n", connect_duration)
		fmt.Fprintf(w, "Download: %s\n", download_duration)
		fmt.Fprintf(w, "Scale: %s\n", scale_duration)
		fmt.Fprintf(w, "Upload: %s\n", upload_duration)

	default:
		fmt.Fprintf(w, "Sorry, only GET and POST methods are supported.")
	}
}

func scaleImages(localPath string, width int, height int) {
	cnt := 0
	filepath.Walk(localPath, func(path string, info fs.FileInfo, err error) error {
		if info.IsDir() {
			return nil
		}

		file, err := os.Open(path)
		if err != nil {
			log.Fatal(err)
		}

		// decode jpeg into image.Image
		img, err := jpeg.Decode(file)
		if err != nil {
			log.Fatal(err)
		}
		file.Close()

		m := resize.Resize(uint(width), uint(height), img, resize.Lanczos3)

		out, err := os.Create(path)
		if err != nil {
			log.Fatal(err)
		}
		defer out.Close()

		// write new image to file
		err = jpeg.Encode(out, m, &jpeg.Options{Quality: 100})
		if err != nil {
			log.Fatal(err)
		}
		cnt++

		return nil
	})
	log.Printf("Scaled %d images\n", cnt)
}

File: python/image-scale/image-scale.py
import os
import shutil
import json
import time
from datetime import date, datetime

import argparse
import uuid
from minio import Minio
from PIL import Image
from flask import Flask, request, make_response

app = Flask(__name__)

parser = argparse.ArgumentParser(
    prog='Image Recognition',
    description='Runs resnet on the images',
)
parser.add_argument('-p', '--port', type=int, default=9090)
parser.add_argument('--storage_path', type=str, default=None)
args = parser.parse_args()

endpoint = "10.121.240.169:9000"
access_key = "LbtKL76UbWedONnd"
secret_key = "Bt0Omfh0S3ud5VEQAVR85CwinSULl3Sj"


def downloadImages(minio_client: Minio, bucket_name, remote_path, local_path):
    obj_lst = minio_client.list_objects(bucket_name, remote_path, False)
    cnt = 0
    for obj in obj_lst:
        minio_client.fget_object(bucket_name, obj.object_name, local_path + os.path.basename(obj.object_name))
        cnt += 1
    obj_lst.close()
    return cnt


def uploadImages(minio_client: Minio, bucket_name, local_path, remote_path):
    for filename in os.listdir(local_path):
        filepath = os.path.join(local_path, filename)
        minio_client.fput_object(bucket_name, remote_path + filename, filepath)
    return


def copyImages(src_path, dst_path):
    if os.path.exists(dst_path):
        shutil.rmtree(dst_path)
    shutil.copytree(src_path, dst_path)


def scaleImages(local_path, width, height):
    for filename in os.listdir(local_path):
        filepath = os.path.join(local_path, filename)
        img = Image.open(filepath)
        img = img.resize((width, height))
        img.save(filepath, 'JPEG')
    return


@app.route('/', methods=['POST'])
def imageRecognition():
    code_start_time = time.perf_counter()

    data = request.data.decode("utf-8")
    data = json.loads(data)

    if args.storage_path is not None:
        bucket_name = data['Bucket'].rstrip("/")
        download_path = data['Source'].rstrip("/") + "/"
        upload_path = args.storage_path.rstrip("/") + "/" + data['Destination'].rstrip("/") + "/"
        local_path = './storage/'
    else:
        bucket_name = data['Bucket'].rstrip("/")
        download_path = data['Source'].rstrip("/") + "/"
        upload_path = data['Destination'].rstrip("/") + "/"
        local_path = './storage/'

    # remove exist storage and create
    if os.path.exists(local_path):
        shutil.rmtree(local_path)
    os.makedirs(local_path)

    minio_client = Minio(
        endpoint,
        access_key=access_key,
        secret_key=secret_key,
        secure=False
    )
    print(f"Connected to {endpoint}")

    download_start_time = time.perf_counter()
    downloadImages(minio_client, bucket_name, download_path, local_path)
    download_end_time = time.perf_counter()
    download_duration = download_end_time - download_start_time

    scale_start_time = time.perf_counter()
    scaleImages(local_path, 224, 224)
    scale_end_time = time.perf_counter()
    scale_duration = scale_end_time - scale_start_time

    upload_start_time = time.perf_counter()
    if args.storage_path is not None:
        copyImages(local_path, upload_path)
    else:
        uploadImages(minio_client, bucket_name, local_path, upload_path)

    upload_end_time = time.perf_counter()
    upload_duration = upload_end_time - upload_start_time

    code_end_time = time.perf_counter()
    code_duration = code_end_time - code_start_time
    print(f"Execution time: {code_duration}")
    print(f"Download time: {download_duration}")
    print(f"Scale time: {scale_duration}")
    print(f"Upload time: {upload_duration}")

    # send response
    response = make_response({})
    response.headers["Content-Type"] = "application/json"
    response.headers["Ce-Id"] = str(uuid.uuid4())
    response.headers["Ce-specversion"] = "0.3"
    response.headers["Ce-Source"] = "image-scale"
    return response


if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=args.port)
